### ISSUE TITLE
Features are always resolved at workspace level

### ISSUE THREAD

**Issue by Shnatsel** [rust-secure-code/cargo-auditable#66]
We use `cargo metadata`, so we are affected by this issue: https://github.com/rust-lang/cargo/issues/7754

**Comment by sunshowers** [comment]
Hey!

I'm currently working through using cargo-auditable and/or cyclonedx to generate an SBOM, and I've run into this general class of issues.

To address them, I'm planning to hook up the Cargo build simulations I wrote within [guppy](https://docs.rs/guppy/0.17.10/guppy/graph/cargo/index.html). These build simulations are, to my knowledge, correct for both the v1 and v2 resolvers, as validated by extensive randomized testing. (They're also the foundation of [cargo-hakari](https://docs.rs/cargo-hakari/latest/cargo_hakari/)).

The idea is that guppy's build simulations will produce the exact set of packages and features that a cargo build would run, even if the build is run on a *host* platform that differs from the current platform.

I'm planning to maintain a fork for now. Would there be interest in upstreaming it?

**Comment by sunshowers** [comment]
I'm also curious what your thoughts are about cargo-auditable vs cargo-cyclonedx. We're currently evaluating which approach to go with (data stored in the binary vs a file on the side).

**Comment by Shnatsel** [comment]
Thank you for reaching out! Yes, this is interesting. I have also been looking into the [krates](https://crates.io/crates/krates) crate, which also lets one write custom traversal functions over the crate graph and replicate resolver v2 fairly easily that way. 

I am cautious to introduce custom reimplementations of complex and loosely documented algorithms, but extensive randomized testing gives me enough confidence in its correctness.

My concern is that Cargo is about to roll out the MSRV-aware [resolver v3](https://github.com/rust-lang/cargo/issues/13873), which is going to ship in rust 1.85 in February 2025. This is the same release that will stabilize the 2024 edition and make MSRV-aware resolver the default. So whatever algorithm we use to refine `cargo metadata` output would also have to replicate resolver v3. 


**TL;DR:** I would be happy to switch to guppy if it also supports the upcoming resolver v3. And I guess we'd still have to maintain a raw `cargo metadata` fallback for future resolver versions.

**Comment by Shnatsel** [comment]
Regarding cargo-auditable vs cargo-cyclonedx: I've worked on both tools. "both have benefits and costs" just about sums it up.

**Comment by sunshowers** [comment]
Thanks for the response!

Resolver v3 only affects the part of the dependency resolver that finds which versions to resolve to. Feature resolution is the same as v2, so no code changes are required. I'll try pushing an update to guppy to add the v3 resolver as an option, which works the same as the v2 resolver.

I would definitely be careful about trying to reimplement the resolver yourself. While building guppy's Cargo resolution I found a number of edge cases. For example:

* the build-dependencies table is only taken into account if a build.rs is present
* proc macros are right at the edge of host and target dependencies, so need to be handled carefully. Guppy provides a couple of knobs to handle a few different scenarios around them.
* To handle weak features, guppy uses a custom spin on a depth-first search called a "buffered filtered DFS".
* There's also all the usual bits folks are familiar with in Cargo, such as the fact that [dependency cycles are permitted](https://docs.rs/guppy/latest/guppy/graph/struct.Cycles.html) due to dev-dependencies.

The resolver lives in
https://github.com/guppy-rs/guppy/blob/main/guppy/src/graph/cargo/build.rs. It's relatively short, but that's only because it's built on a lot of the more generic scaffolding that guppy provides.

Interestingly, in comparison testing I've found that guppy is many times faster than Cargo at delivering the same results. It's bad enough that guppy's CI runs the comparison tests in release mode — guppy itself is plenty fast in debug mode (even with opt-level 1), but Cargo is too slow when you're running hundreds of simulations per run.

**Comment by Shnatsel** [comment]
Curiously, Cargo itself is in the process of migrating to PubGrub to speed things up, and also make the dependency resolver a standalone crate: https://rust-lang.github.io/rust-project-goals/2024h2/pubgrub-in-cargo.html

The tracking issue is [here](https://github.com/rust-lang/rust-project-goals/issues/110), and they have also demonstrated a significant performance improvement - see the [comment from 3 days ago](https://github.com/rust-lang/rust-project-goals/issues/110#issuecomment-2552307967). 

If I add some dependency solver, I'd prefer it to be that one, because this will be identical to what Cargo itself uses. That would get us if not perfect correctness then at least perfect adherence to Cargo's quirks. Do you think it would be a good idea to just use that in place of guppy?

**Comment by sunshowers** [comment]
Hmm I thought pubgrub was just for version resolution, not feature resolution (which isn't a constraint solving problem)

**Comment by Shnatsel** [comment]
I may be very wrong here! Let me ask the person working on PubGrub in Cargo.

**Comment by Shnatsel** [comment]
Indeed, the bits to be exposed by Cargo are [not sufficient](https://rust-lang.zulipchat.com/#narrow/channel/260232-t-cargo.2FPubGrub/topic/Progress.20report/near/490339607) for the task.

Then `guppy` is clearly the best option, at least until Cargo adds [native SBOM precursors](https://github.com/rust-lang/cargo/pull/13709) and it's not clear if that's ever going to happen.

I'll be happy to add `guppy` to `cargo auditable` once resolver v3 is supported. We'll still need to keep the fallback `cargo metadata` codepath around for future/unrecognized resolver versions, but I'm excited to finally be able to handle resolvers v2 and v3 properly!

**Comment by sunshowers** [comment]
Resolver v3 is now supported as of guppy 0.17.11. See https://docs.rs/guppy/latest/guppy/graph/cargo/enum.CargoResolverVersion.html#variant.V3.

**Comment by sunshowers** [comment]
> Thank you! Is there any example code I could use as a starting point for integrating `guppy`? Say, could you point me to how cargo-workspace-hack uses it?

This code can help you get started: https://github.com/guppy-rs/guppy/blob/6769bd5748b5a0cfe243db2a20b95a41ccc14878/cargo-guppy/src/lib.rs#L213-L296

Hakari (the workspace-hack manager) does a few things as well. See https://github.com/guppy-rs/guppy/blob/6769bd5748b5a0cfe243db2a20b95a41ccc14878/tools/hakari/src/hakari.rs#L988-L1007, and in general search for `CargoOptions` in that file.

> I'm particularly interested in how the resolver version is determined.

The resolver version actually must be passed in -- it is not auto-determined. One of guppy's design principles is that it works purely in-memory with information passed into it (typically the output of `cargo metadata`), without accessing the environment in any way. (There were at some point plans to use guppy on the web targeting wasm, so it is built to enable that.)

Unfortunately, `cargo metadata` doesn't contain the resolver version, so it must be passed in separately via `CargoResolverVersion`. I wish cargo made it available in a machine-readable form, but until then I think the following algorithm would work with current versions of Cargo.

1. Find the workspace root and read its `Cargo.toml`. Guppy can help with this: https://docs.rs/guppy/latest/guppy/graph/struct.Workspace.html#method.root returns the root directory.
2. If it has `workspace.resolver`, use that.
3. Otherwise, if it has `package.edition`, use that (including a consideration for `package.edition.workspace`).
4. Otherwise, use resolver v1.

Maybe there's a crate already which does this discovery.

**Comment by sunshowers** [comment]
Oh, and the first example linked also shows how to set a different host platform. This feature allows hakari to work in a completely platform-independent manner, such that the same `workspace-hack/Cargo.toml` file is produced no matter what platform you're on. This is something that Cargo's current APIs don't enable.

**Comment by Shnatsel** [comment]
I [checked](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/inspecting.20Cargo.20resolver.20version/near/495157543) with the Cargo team, and obtaining the resolver version without Cargo directly exposing it seems to be practical!

I am now convinced that using guppy is feasible. I'd appreciate a pull request for migrating cargo-auditable to guppy. Otherwise I'll try to get around to writing it myself in the next month or so.

**Comment by sunshowers** [comment]
Gotcha, thanks for checking! I think "get the resolver" is a useful primitive that can maybe be made into a separate crate (part of guppy maybe).

I think at the moment, at Oxide we're leaning towards building on top of `cargo cyclonedx` rather than using `cargo auditable` (for several reasons, among which is the fact that many of our artifacts consist of lots of binary and non-binary components). What are your thoughts on that?

**Comment by Shnatsel** [comment]
Or you could use both tools. This will let you list as many non-binary artifacts as you like in CycloneDX yet keep the binaries-self describing and make their SBOMs impossible to misplace.

**Comment by Shnatsel** [comment]
Regarding the underlying implementation, I would prefer to rely on information exposed by Cargo rather than compatible reimplementations in the long term. 

Cargo has a project goal for the first half of this year to prototype better plumbing subcommands to expose Cargo information to the outside world and bypass the need for third-party reimplementations: https://rust-lang.github.io/rust-project-goals/2025h1/cargo-plumbing.html

We tried this earlier with https://github.com/rust-lang/cargo/pull/13709 but that never got merged. Now that it's a project goal and has a Cargo maintainer responsible for it, such work should be a lot easier to land. However, right now they don't have anyone to actually do the brunt of the work.

**Comment by sunshowers** [comment]
Thanks -- yes, absolutely agreed about querying Cargo being the right long-term direction. I'll be honest when I started writing guppy in 2019, I was quite surprised that information wasn't already available!

**Comment by Shnatsel** [comment]
I've started working on a standalone crate to get a Cargo package version: https://github.com/Shnatsel/resolverver

**Comment by Shnatsel** [comment]
`resolverver` crate MVP is now usable end-to-end. It's still missing documentation, and could use more testing on real-world data. I'm not sure how to arrange said testing - there doesn't seem to be a way to get Cargo to print the resolver version, so it's hard to get the ground truth.

Would you like to fold the code into the guppy repo, or keep it standalone? Also, would you like me to publish the crate to crates.io, or would you rather keep that to yourself?

**Comment by sunshowers** [comment]
Wow that's awesome! Do you feel up for maintaining it? If so I would prefer that :) otherwise we can move it into the guppy repo, or a separate repo in the guppy org. What do you think?

**Comment by Shnatsel** [comment]
It's already basically done, so I don't think it will require much maintenance. So sure, I'm up for it. I can simply import it into cargo-auditable workspace and publish it to crates.io myself.

**Comment by Shnatsel** [comment]
Alright, I added docs to `resolverver`, moved the crate into this workspace, and reserved the name on crates.io.

@sunshowers are you still interested in contributing a PR to migrate `cargo auditable` to `guppy` now that the last missing piece is in place?

**Comment by sunshowers** [comment]
@Shnatsel probably not for another week at least. Been really busy with higher priority things at the moment.

**Comment by Shnatsel** [comment]
No worries, it isn't urgent. I'm glad this work is no longer blocked by any technical limitations at least.

**Comment by jayvdb** [comment]
An approach that I have been investigating seems to work.

Currently it requires that `target/$PROFILE/deps` is empty before starting.

Then do a build , and read all of the generated .d files in that directory and extract the <crate>-<version> from each.  This list is very close to the list of compiled crates that `cargo build` lists.

The missing items not in the .d files are workspace members.

I have desk checked this on a rather large project, and combined those two form the correct set of dependencies.  I would love to know if there are any problems with this algorithm, beside the obvious ugliness of it.

It seems like all this can be done within `encode_audit_data` , filtering the `packages` and `packages[x].dependencies` that to remove crates which were not compiled, and this would go into the very end of the function before it returns the reduced set at https://github.com/rust-secure-code/cargo-auditable/blob/ed48186/cargo-auditable/src/auditable_from_metadata.rs#L208

It would be lovely if anyone knows how to find the subset of .d files included in the build, so that `target/$PROFILE/deps` doesnt need to be deleted prior to the build.

**Comment by Shnatsel** [comment]
That is a clever hack, but unfortunately it runs into the exact same problem when workspaces are involved: if you `cargo build`  at the root of a workspace, it will build all workspace crates, so they all will write .d files, so you get the dependencies from the entire workspace merged, and we're back to square one. The only process that actually yields an accurate SBOM is building each workspace crate separately and running `cargo clean` after each build, but that entirely defeats the point of having a workspace.

#192 would resolve this robustly and I'm hoping to implement that later this year, but that will be nightly-only for a while, potentially years.

It would be nice to have a more accurate fallback on stable. `guppy` seems to be a good option, at the cost of rather high complexity. But if we're desperate enough to resort to parsing lists that were never meant to be parsed, `cargo tree` might be a viable source of truth.

**Comment by Shnatsel** [comment]
I've opened https://github.com/rust-secure-code/cargo-auditable/pull/210 which should work around the limitations of `cargo metadata` by parsing the output of `cargo tree` and cross-referencing the package list against that.

Please try it and let me know if this fixes the issue for you,
